### ISSUE TITLE
Supporting input-specific baselines

### ISSUE THREAD

**Issue by jmschrei** [pytorch/captum#933]
## 🚀 Feature
Currently, it seems like all baseline examples are used for each input. Would it be possible, potentially by adding support for passing in a tensor with one more dimension than the shape of the input tensor, to pass in input-specific baselines?

## Motivation
Some of my work involves shuffling each input sequence to use as a sequence-specific baseline Since this has to be done separately for each input example, I have to run DeepLiftShap iteratively one sequence at a time. I imagine that this doesn't fully utilize the GPU. 

## Pitch
GPU go brrrr. 

## Alternatives
Perhaps the existing code could be used if there was a wrapper or something that asynchronously submitted all the sequences, one at a time?

**Comment by NarineK** [comment]
@jmschrei, currently we can specify an example specific baseline for each example in the batch (this is the first dimension of the inputs and baselines). If the inputs and baselines have the same shape then we are taking example specific baseline from the batch for most attribution methods(e.g. integrated gradients; https://github.com/pytorch/captum/blob/master/captum/attr/_core/integrated_gradients.py#L145) 
DeepLiftShap specifically is looking into a distribution of baselines per example that's why it considers all the baselines provided in the baselines argument. We could potentially  add an additional dimension for more than one baseline distribution but that will be API inconsistency and users need to understand why it is different for DeepLiftShap

**Comment by jmschrei** [comment]
Hi @NarineK 

That makes sense to me. I'm not sure that I see the option of having a baseline tensors with an additional tensor as an "inconsistency," because users could continue to use the code as-is without behaviour changes. 

I can see the argument for not making the change, though. Do you know any workarounds that I can use to run something like this on my end without needing to change the code? Thanks!

**Comment by NarineK** [comment]
@jmschrei, DeepLiftShap is a simple wrapper on top of DeepLift that expands the baselines and inputs accordingly and calls Deeplift. You could potentially Implement your own wrapper of DeepLiftShap that adjusts inputs and baselines as expected and calls DeepLift.
https://github.com/pytorch/captum/blob/master/captum/attr/_core/deep_lift.py#L853

**Comment by yztxwd** [comment]
I recently ran into the same issue that I need to have example-specifc background sets (also by dinucleotide shuffling), I did that by creating the pool of all background sets for all examples in the batch first, then repeat_interleave the input examples to match the background pool. I just write it down for anyone interested in doing something similar

Here's the details:

1. I define a baselines function that takes a batch of examples and prepares a background set pool in a way like:
background 1 for example 1
background 2 for example 1
...
background 10 for example 1
background 1 for example 2
...
background 10 for example n

2. Then I created a new DeepLiftShapPatched class which overwrites the `_expand_inputs_baselines_targets` and `_compute_mean_across_baselines` to align the repeat_interleaved examples to the backgrounds

```python
    def _expand_inputs_baselines_targets(
        self,
        baselines: Tuple[Tensor, ...],
        inputs: Tuple[Tensor, ...],
        target: TargetType,
        additional_forward_args: Any,
        num_baselines_per_sample: int=10,
    ) -> Tuple[Tuple[Tensor, ...], Tuple[Tensor, ...], TargetType, Any]:

        inp_bsz = inputs[0].shape[0]
        #base_bsz = baselines[0].shape[0]
        base_bsz = num_baselines_per_sample

        expanded_inputs = tuple(
            [
                input.repeat_interleave(base_bsz, dim=0).requires_grad_()
                for input in inputs
            ]
        )
        # no need to repeat baselines here, we already did it in the baseline function
        #expanded_baselines = tuple(
        #    [
        #        baseline.repeat(
        #            (inp_bsz,) + tuple([1] * (len(baseline.shape) - 1))
        #        ).requires_grad_()
        #        for baseline in baselines
        #    ]
        #)
        expanded_baselines = baselines
        expanded_target = _expand_target(
            target, base_bsz, expansion_type=ExpansionTypes.repeat_interleave
        )
        input_additional_args = (
            _expand_additional_forward_args(
                additional_forward_args,
                base_bsz,
                expansion_type=ExpansionTypes.repeat_interleave,
            )
            if additional_forward_args is not None
            else None
        )
        return (
            expanded_inputs,
            expanded_baselines,
            expanded_target,
            input_additional_args,
        )

    def _compute_mean_across_baselines(
        self, inp_bsz: int, base_bsz: int, attribution: Tensor, num_baselines_per_sample: int=10
    ) -> Tensor:
        # Average for multiple references
        attr_shape: Tuple = (inp_bsz, num_baselines_per_sample)
        if len(attribution.shape) > 1:
            attr_shape += attribution.shape[1:]
        return torch.mean(attribution.view(attr_shape), dim=1, keepdim=False)
```

I added another parameter `num_baselines_per_sample` is because I know there are exactly 10 backgrounds for each example, as I did in the baselines function. The inputs are repeat_interleaved 10 times to match that, and the baselines are kept the same. Also in the `_compute_mean_across_baselines` I just take the mean of every 10 attributions. Hope that makes sense
